homeassistant/components/firmata/binary_sensor.py

```diff
@@ -0,0 +1,103 @@
+"""Support for Firmata binary sensor input."""
+
+import logging
+
+from homeassistant.components.binary_sensor import BinarySensorEntity
+from homeassistant.core import callback
+
+from .board import FirmataBoardPin
+from .const import CONF_NEGATE_STATE, DOMAIN, PIN_MODE_INPUT, PIN_MODE_PULLUP
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the Firmata binary sensors."""
+    new_entities = []
+
+    boards = hass.data[DOMAIN]
+    board = boards[config_entry.entry_id]
+    for binary_sensor in board.binary_sensors:
+        binary_sensor_entity = FirmataDigitalBinaryInput(
+            hass, config_entry, **binary_sensor
+        )
+        new_binary_sensor = await binary_sensor_entity.setup_pin()
+        if new_binary_sensor:
+            new_entities.append(binary_sensor_entity)
+
+    async_add_entities(new_entities)
+
+
+class FirmataDigitalBinaryInput(FirmataBoardPin, BinarySensorEntity):
+    """Representation of a Firmata Digital Input Pin."""
+
+    async def setup_pin(self):
+        """Set up a digital input pin."""
+        _LOGGER.debug(
+            "Setting up binary sensor pin %s for board %s", self._name, self._board_name
+        )
+        if not self._mark_pin_used():
+            _LOGGER.warning(
+                "Pin %s already used! Cannot use for binary sensor %s",
+                self._pin,
+                self._name,
+            )
+            return False
+        api = self._board.api
+        if self._pin_mode == PIN_MODE_INPUT:
+            await api.set_pin_mode_digital_input(self._pin, self.latch_callback)
```

Comment:
The entity callback should not be registered before the entity has been added to home assistant. We can use `async_added_to_hass` for this purpose.

If we need to set up the pin before adding the entity we should use our dispatch helper to signal the entity callback from the pin callback.